src/libbson/src/bson/bson-utf8.c

```diff
@@ -264,77 +376,104 @@ bson_utf8_escape_for_json (const char *utf8, /* IN */
+            bson_string_append_ex (str, utf8, pos);
+            break;
+         }
+
+         continue;
+      }
+
+      // Reached a special character. Copy over all of normal characters
+      // we have passed so far
+      if (pos) {
+         bson_string_append_ex (str, utf8, pos);
+         utf8 += pos;
+         pos = 0;
+      }
+
       // Check if expected char length goes past end
+      // bson_utf8_get_char will crash without this check
       _bson_utf8_get_sequence (utf8, &length_of_char, &mask);
       if (utf8 > end - length_of_char) {
          goto invalid_utf8;
       }
 
-      c = bson_utf8_get_char (utf8);
+      // Check for null character
+      // Null characters are only allowed if the length is provided
+      if (!*utf8 || (*utf8 == '\xc0' && *(utf8 + 1) == '\x80')) {
+         if (!length_provided) {
+            goto invalid_utf8;
+         }
 
-      switch (c) {
-      case '\\':
-      case '"':
-         bson_string_append_c (str, '\\');
-         bson_string_append_unichar (str, c);
-         break;
-      case '\b':
-         bson_string_append (str, "\\b");
-         break;
-      case '\f':
-         bson_string_append (str, "\\f");
-         break;
-      case '\n':
-         bson_string_append (str, "\\n");
-         break;
-      case '\r':
-         bson_string_append (str, "\\r");
-         break;
-      case '\t':
-         bson_string_append (str, "\\t");
-         break;
-      default:
-         if (c < ' ') {
-            bson_string_append_printf (str, "\\u%04x", (unsigned) c);
-         } else {
-            bson_string_append_unichar (str, c);
+         bson_string_append (str, "\\u0000");
+         utf8_len -= *utf8 ? 2 : 1;
+         utf8 += *utf8 ? 2 : 1;
+         continue;
+      }
+
+      // Unicode character
+      if (c > 0x7f /* highest ASCII character */) {
+         const char *utf8_old = utf8;
+         uint8_t char_len;
+
+         c = bson_utf8_get_char (utf8);
+
+         if (!c) {
+            goto invalid_utf8;
          }
-         break;
+
+         bson_string_append_unichar (str, c);
+         utf8 = bson_utf8_next_char (utf8);
+
+         char_len = utf8 - utf8_old;
+         utf8_len -= char_len;
+         pos = 0;
+
+         continue;
       }
 
+      // Special ASCII characters (control chars and misc.)
+      _bson_utf8_handle_special_char (c, str);
+
       if (c) {
-         utf8 = bson_utf8_next_char (utf8);
+         utf8++;
       } else {
-         if (length_provided) {
-            // Check if current character is null character,
-            // if so skip past it as we've already added '\\u0000' above
-            if (*utf8 == '\0' || (*utf8 == '\xc0' && *(utf8 + 1) == '\x80')) {
-               utf8 += (*utf8 == '\0') ? 1 : 2;
-            } else {
-               goto invalid_utf8;
-            }
-         } else {
-            goto invalid_utf8;
-         }
+         goto invalid_utf8;
       }
-   }
+
+      utf8_len--;
+   } while (utf8_len);
```

Comment:
```suggestion
   } while (utf8_len > 0);
```
Avoid treating integers as booleans.